search.go

```diff
@@ -0,0 +1,213 @@
+	Code    string `json:"code"`
+	Message string `json:"message"`
+}
+
+// SearchJobStates are the different states a search job can be in.
+var SearchJobStates = map[string]string{
+	"NOT STARTED":            "Search job has not been started yet.",
+	"GATHERING RESULTS":      "Search job is still gathering more results, however results might already be available.",
+	"FORCE PAUSED":           "Query that is paused by the system. It is true only for non-aggregate queries that are paused at the limit of 100k. This limit is dynamic and may vary from customer to customer.",
+	"DONE GATHERING RESULTS": "Search job is done gathering results; the entire specified time range has been covered.",
+	"CANCELED":               "The search job has been canceled.",
+}
+
+// StartSearch calls the Sumologic API Search Endpoint.
+// POST search/jobs
+func (c *Client) StartSearch(ssr StartSearchRequest) (*SearchJob, []*http.Cookie, error) {
+	body, _ := json.Marshal(ssr)
+
+	relativeURL, _ := url.Parse("search/jobs")
+	url := c.EndpointURL.ResolveReference(relativeURL)
+
+	req, err := http.NewRequest("POST", url.String(), bytes.NewBuffer(body))
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Authorization", "Basic "+c.AuthToken)
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer resp.Body.Close()
+
+	responseBody, _ := ioutil.ReadAll(resp.Body)
+
+	switch resp.StatusCode {
+	case http.StatusAccepted:
+		var sj = new(SearchJob)
+
+		err = json.Unmarshal(responseBody, &sj)
+		if err != nil {
+			return nil, nil, err
+		}
+		cookies := resp.Cookies()
+		return sj, cookies, nil
+	case http.StatusUnauthorized:
+		return nil, nil, ErrClientAuthenticationError
+	case http.StatusBadRequest:
+		var sj = new(SearchJob)
+		err = json.Unmarshal(responseBody, &sj)
+		if err != nil {
+			return nil, nil, err
+		}
+		return nil, nil, fmt.Errorf("Start SearchJob BadRequest, %v, %v", sj.Code, sj.Message)
+	default:
+		return nil, nil, fmt.Errorf("unexepected http status code %v", resp.StatusCode)
+	}
+}
+
+// HistogramBucket corresponds to the histogram display in the Sumo Logic interactive analytics API.
+type HistogramBucket struct {
+	Length         int `json:"length"`
+	Count          int `json:"count"`
+	StartTimeStamp int `json:"startTimeStamp"`
+}
+
+// SearchJobStatusResponse stores the response from getting a search status.
+type SearchJobStatusResponse struct {
+	State           string             `json:"state"`
+	MessageCount    int                `json:"messageCount"`
+	HistgramBuckets []*HistogramBucket `json:"histogramBuckets"`
+	RecordCount     int                `json:"recordCount"`
+	PendingWarnings []string           `json:"pendingWarnings"`
+	PendingErrors   []string           `json:"pendingErrors"`
+}
+
+// GetSearchJobStatus retrieves the status of a running job.
+func (c *Client) GetSearchJobStatus(searchJobID string, cookies []*http.Cookie) (*SearchJobStatusResponse, error) {
```

Comment:
If cookies were a private field on SearchJob, this method could be implemented as `SearchJob{}.Status()`